File: math/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int carry = 0;
        ListNode* head = new ListNode(0);
        ListNode* cur = head;
        while(l1!=NULL || l2!=NULL || carry!=0){
            int x = 0;
            if(l1!=NULL){
                x =l1->val;
                l1=l1->next;
            }

            int y = 0;
            if(l2!=NULL){
                y = l2->val;
                l2=l2->next;
            }
            int sum = x+y+carry;
            carry = sum/10;
            int val = sum%10;
            ListNode* newNode = new ListNode(val);
            cur->next = newNode;
            cur = cur->next;

            
           
        }

        return head->next;
    }
};

File: recursion/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        // Handle l1
        ListNode cur1 = l1, cur2 = l2; 
        long digit = 1; 
        long sum = 0; 
        while(cur1 != null){
            sum += cur1.val * digit;
            digit *= 10; 
            cur1 = cur1.next; 
        }
        // Handle l2
        digit = 1; 
        while(cur2 != null){
            sum += cur2.val * digit;
            digit *= 10;  
            cur2 = cur2.next; 
        }
        if (sum == 0){
            return new ListNode(0); 
        }

        // Organize return
        ListNode dummy = new ListNode(); 
        ListNode cur = dummy; 
        while(sum != 0){
            int curDigit = (int)sum % 10; 
            sum /= 10; 
            ListNode curNode = new ListNode(curDigit); 
            cur.next = curNode; 
            cur = cur.next; 
        }
        return dummy.next; 
    }
}